### DISCUSSION TITLE
Can I increase the useDerivedState hook?

### DISCUSSION THREAD

**Discussion by vaynevayne** [mantinedev/mantine#3890] [Feature requests]
Encapsulating intermediate components: used when allowing controlled components to support uncontrolled calls

```jsx
import { Dispatch, SetStateAction, useState } from "react";

// 这段代码定义了一个自定义hook，用于派生状态
const useDerivedState = <T = any>(
  value: T
): [T, Dispatch<SetStateAction<T>>] => {
  // 使用useState定义内部状态和前一个状态
  const [internalValue, setInternalValue] = useState<T>(value);
  const [prevValue, setPrevValue] = useState<T>(value);

  // 如果传入的value和前一个状态不同，则更新内部状态和前一个状态
  if (value !== prevValue) {
    setPrevValue(value);
    setInternalValue(value);
  }
  // 返回内部状态和更新内部状态的函数
  return [internalValue, setInternalValue];
};

export default useDerivedState;

```

```jsx
const LazyButton: FC<LazyButtonProps> = ({
  value,
  defaultValue,
  onChange,
  ...rest
}) => {
  const [opened, { open, close }] = useDisclosure(false)

  const [internalValue, setInternalValue] = useDerivedState(value)

  return (
   
        <Modal opened={opened} onClose={close} title={label}>
          <SomeConrolledComponent      value={internalValue}
              onChange={setInternalValue}
              defaultValue={defaultValue}
              {...rest}/>
        
          <Group position="right">
            <Button
              type="submit"
              onClick={() => {
                onChange?.(internalValue)
                close()
              }}
            >
              确定
            </Button>
          </Group>
        </Modal>

  )
}
```

Usage 
```jsx
  <LazyModal
                value={columnDefs}
                onChange={setColumnDefs}
              />
```

**Comment by rtivital** [comment]
See use-uncontrolled hook, it does something similar – https://mantine.dev/hooks/use-uncontrolled/

**Reply by vaynevayne** [reply]
Not the same,

```jsx
const App = ({value,onChange}) = > {
const [_value,_onChange] = useUncontrolled({
value,
onChange
})
}
```

When _ value changed, cause emit onChange immediate, but sometimes, we need to delay calling onChange. so we need derived a new state used internal,  For example, when Modal is closed, the value is passed to the parent component, which une-uncontrolled cannot do.

Also, defaultValue is only used once for initialization. props cannot be synchronized. At some point, we need to sync props

**Reply by vaynevayne** [reply]
namely 

use-uncontrolled support 
```jsx
// 1. 
useUncontrolled({
value,
onChange
})

// 2. 
useUncontrolled({
defaultValue
})

```
not support , it is useDerivedState does

```jsx
// not support
useUncontrolled({
value
})
```

useUncontrolled's state is always on props, so onChange must be called when the value is changed. If you don't want to trigger onChange, use useDerivedState.

**Comment by vaynevayne** [comment]
Actually, I think there's something wrong with the `use-uncontrolled` naming, he's supposed to be use-controlled,
Chakra-ui also has this hook to do the same thing, but his name is `use-controllableState`.

`UseDerivedState` is the use-uncontrolled.

https://chakra-ui.com/docs/hooks/use-controllable
